### PR TITLE
Tableau

### PR DESCRIPTION
**Pull request by ityshko**

### CODE REVIEW

File: CdapOdbc/ColumnInfo.cpp [hunk 1]
```diff
@@ -23,4 +23,13 @@
 Cask::CdapOdbc::ColumnInfo::ColumnInfo(const ColumnDesc& columnDesc)
   : name(columnDesc.getName())
   , dataType(Driver::getInstance().getDataType(columnDesc.getType())) {
+
+  std::wstring::size_type found;
+  if ((found = this->name.rfind(L".", this->name.length() - 1)) != std::wstring::npos) {
```
Comment (skilurus) [inline CdapOdbc/ColumnInfo.cpp]: Move expression to to the line above. Don't like complex expressions inside IF, especially assignments.

File: CdapOdbc/ColumnInfo.cpp [hunk 2]
```diff
@@ -23,4 +23,13 @@
 Cask::CdapOdbc::ColumnInfo::ColumnInfo(const ColumnDesc& columnDesc)
   : name(columnDesc.getName())
   , dataType(Driver::getInstance().getDataType(columnDesc.getType())) {
+
+  std::wstring::size_type found = this->name.rfind(L".", this->name.length() - 1);
+  if (found != std::wstring::npos) {
+    this->shortName = columnDesc.getName().substr(found + 1);
+  }
+  else {
```
Comment (ityshko) [inline CdapOdbc/ColumnInfo.cpp]: style: move 'else' up

File: CdapOdbc/DataReader.cpp [hunk 3]
```diff
@@ -170,6 +182,29 @@ void Cask::CdapOdbc::DataReader::fetchValue(const web::json::value& value, const
...
+
+      this->fetchUnsignedLong(ubintValue, binding);
+      break;
+
+    case SQL_C_SBIGINT:
+      if (value.is_string()) {
+        sbintValue = std::wcstol(value.as_string().c_str(), nullptr, 0);
+      } else if (value.is_number()) {
+        sbintValue = value.as_number().to_int64();
+      }	else if (value.is_boolean()) {
```
Comment (ityshko) [inline CdapOdbc/DataReader.cpp]: extra tab

File: CdapOdbc/DataReader.h [hunk 4]
```diff
@@ -38,7 +38,9 @@ namespace Cask {
       void fetchWVarchar(const wchar_t* str, const ColumnBinding& binding);
       void fetchInt(SQLINTEGER value, const ColumnBinding& binding);
       void fetchSmallint(SQLSMALLINT value, const ColumnBinding& binding);
-      void fetchDouble(SQLDOUBLE value, const ColumnBinding& binding);
+			void fetchUnsignedLong(SQLUBIGINT value, const ColumnBinding& binding);
```
Comment (ityshko) [inline CdapOdbc/DataReader.h]: extra tabs

File: CdapOdbc/NoSchemaColumnsDataReader.cpp [hunk 5]
```diff
@@ -41,18 +41,33 @@ void Cask::CdapOdbc::NoSchemaColumnsDataReader::getColumnValue(const ColumnBindi
     case 16: // CHAR_OCTET_LENGTH 
       this->fetchNull(binding);
       break;
-    case 3: // TABLE_NAME 
-      this->fetchVarchar(this->tableName.c_str(), binding);
+    case 3: // TABLE_NAME
+      if (binding.getTargetType() == SQL_WCHAR) {
+        this->fetchWVarchar(this->tableName.c_str(), binding);
+      }
+      else {
```
Comment (ityshko) [inline CdapOdbc/NoSchemaColumnsDataReader.cpp]: move 'else' up

File: CdapOdbc/NoSchemaColumnsDataReader.cpp [hunk 6]
```diff
@@ -41,18 +41,33 @@ void Cask::CdapOdbc::NoSchemaColumnsDataReader::getColumnValue(const ColumnBindi
...
+        this->fetchVarchar(this->tableName.c_str(), binding);
+      }
       break;
     case 4: // COLUMN_NAME 
-      this->fetchVarchar(names[this->currentRowIndex], binding);
+			if (binding.getTargetType() == SQL_WCHAR) {
+				this->fetchWVarchar(names[this->currentRowIndex], binding);
+			}
+			else {
```
Comment (ityshko) [inline CdapOdbc/NoSchemaColumnsDataReader.cpp]: move else up

File: CdapOdbc/NoSchemaColumnsDataReader.cpp [hunk 7]
```diff
@@ -41,18 +41,33 @@ void Cask::CdapOdbc::NoSchemaColumnsDataReader::getColumnValue(const ColumnBindi
...
     case 14: // SQL_DATA_TYPE 
       this->fetchSmallint(SQL_CHAR, binding);
       break;
-    case 6: // TYPE_NAME 
-      this->fetchVarchar(L"string", binding);
+    case 6: // TYPE_NAME
+      if (binding.getTargetType() == SQL_WCHAR) {
+        this->fetchWVarchar(L"string", binding);
+      }
+      else {
```
Comment (ityshko) [inline CdapOdbc/NoSchemaColumnsDataReader.cpp]: move 'else' up

File: CdapOdbc/NoSchemaColumnsDataReader.cpp [hunk 8]
```diff
@@ -70,7 +85,12 @@ void Cask::CdapOdbc::NoSchemaColumnsDataReader::getColumnValue(const ColumnBindi
       this->fetchInt(this->currentRowIndex + 1, binding);
       break;
     case 18: // IS_NULLABLE 
-      this->fetchVarchar(L"YES", binding);
+      if (binding.getTargetType() == SQL_WCHAR) {
+        this->fetchWVarchar(L"YES", binding);
+      }
+      else {
```
Comment (ityshko) [inline CdapOdbc/NoSchemaColumnsDataReader.cpp]: move 'else' up

File: CdapOdbc/TablesDataReader.cpp [hunk 9]
```diff
@@ -73,10 +73,19 @@ void Cask::CdapOdbc::TablesDataReader::getColumnValue(const ColumnBinding& bindi
       break;
     case 3: // TABLE_NAME 
       name = this->getTableName();
-      this->fetchVarchar(name.c_str(), binding);
+			if (binding.getTargetType() == SQL_WCHAR) {
```
Comment (ityshko) [inline CdapOdbc/TablesDataReader.cpp]: extra tabs

File: CdapOdbc/TablesDataReader.cpp [hunk 10]
```diff
@@ -73,10 +73,19 @@ void Cask::CdapOdbc::TablesDataReader::getColumnValue(const ColumnBinding& bindi
...
+				this->fetchVarchar(name.c_str(), binding);
+			}
       break;
-    case 4: // TABLE_TYPE 
-      this->fetchVarchar(L"TABLE", binding);
+    case 4: // TABLE_TYPE
+      if (binding.getTargetType() == SQL_WCHAR) {
+        this->fetchWVarchar(L"TABLE", binding);
+      }
+      else {
```
Comment (ityshko) [inline CdapOdbc/TablesDataReader.cpp]: move else up

File: CdapOdbc/core.cpp [hunk 11]
```diff
@@ -349,7 +349,27 @@ SQLRETURN SQL_API SQLGetInfoW(
...
+        return SQL_SUCCESS;
+      case SQL_ODBC_INTERFACE_CONFORMANCE:
+        *(reinterpret_cast<SQLUBIGINT*>(InfoValuePtr)) = SQL_OIC_CORE;
+        TRACE(L"SQLGetInfoW returns SQL_SUCCESS, *InfoValuePtr = 1UL\n");
+        return SQL_SUCCESS;
+      case SQL_SQL_CONFORMANCE:
+        *(reinterpret_cast<SQLUBIGINT*>(InfoValuePtr)) = SQL_SC_SQL92_ENTRY;
+        TRACE(L"SQLGetInfoW returns SQL_SUCCESS, *InfoValuePtr = 1UL\n");
+        return SQL_SUCCESS;
+		}
```
Comment (ityshko) [inline CdapOdbc/core.cpp]: extra tabs

File: CdapOdbc/ExploreClient.cpp [hunk 12]
```diff
@@ -189,7 +189,12 @@ QueryResult Cask::CdapOdbc::ExploreClient::getDatasetFields(const std::wstring&
 
 QueryHandle Cask::CdapOdbc::ExploreClient::execute(const std::wstring& statement) {
   web::json::value query;
-  query[L"query"] = toJson(&statement);
+  std::wstring stmt_preproc = statement;
+	// TODO: Remove the wrapping SELECT properly
```
Comment (skilurus) [inline CdapOdbc/ExploreClient.cpp]: Remove unnecessary commented code

File: CdapOdbc/ExploreClient.cpp [hunk 13]
```diff
@@ -189,7 +189,12 @@ QueryResult Cask::CdapOdbc::ExploreClient::getDatasetFields(const std::wstring&
 
 QueryHandle Cask::CdapOdbc::ExploreClient::execute(const std::wstring& statement) {
   web::json::value query;
-  query[L"query"] = toJson(&statement);
+  std::wstring stmt_preproc = statement;
```
Comment (skilurus) [inline CdapOdbc/ExploreClient.cpp]: Remove unnecessary local variable

File: CdapOdbc/Statement.cpp [hunk 14]
```diff
@@ -67,9 +67,9 @@ void Cask::CdapOdbc::Statement::addColumnBinding(const ColumnBinding& binding) {
 }
 
 void Cask::CdapOdbc::Statement::removeColumnBinding(SQLUSMALLINT columnNumber) {
-  if (this->state == State::CLOSED) {
-    this->throwStateError();
-  }
+  //if (this->state == State::CLOSED) {
```
Comment (skilurus) [inline CdapOdbc/Statement.cpp]: Remove commented code

File: CdapOdbc/ExploreClient.cpp (hunk 13 above)
Comment (ityshko) [inline CdapOdbc/ExploreClient.cpp]: not a problem anymore since the variable is used now

File: CdapOdbc/core.cpp [hunk 15]
```diff
@@ -305,7 +304,8 @@ SQLRETURN SQL_API SQLGetInfoW(
         TRACE(L"SQLGetInfoW returns SQL_SUCCESS, InfoValuePtr = %s\n", static_cast<SQLWCHAR*>(InfoValuePtr));
         return SQL_SUCCESS;
       case SQL_IDENTIFIER_QUOTE_CHAR:
-        Argument::fromStdString(L"\"", static_cast<SQLWCHAR*>(InfoValuePtr), BufferLength, StringLengthPtr);
+        Argument::fromStdString(L"", static_cast<SQLWCHAR*>(InfoValuePtr), BufferLength, StringLengthPtr);
+//        Argument::fromStdString(L"\"", static_cast<SQLWCHAR*>(InfoValuePtr), BufferLength, StringLengthPtr);
```
Comment (ityshko) [inline CdapOdbc/core.cpp]: remove unused code

File: CdapOdbc/core.cpp [hunk 16]
```diff
@@ -699,13 +737,141 @@ SQLRETURN SQL_API SQLColAttributeW(
...
+          return SQL_SUCCESS;
+        }
+      case SQL_COLUMN_TYPE_NAME:
+        if (CharacterAttributePtr) {
+          Argument::fromStdString(
+            columnInfo->getDataType().getName(),
+            static_cast<SQLWCHAR*>(CharacterAttributePtr),
+            BufferLength,
+            StringLengthPtr
+            );
```
Comment (ityshko) [inline CdapOdbc/core.cpp]: closing brace must have the same indent as a statement

File: CdapOdbc/core.cpp [hunk 17]
```diff
@@ -699,13 +737,141 @@ SQLRETURN SQL_API SQLColAttributeW(
...
+          TRACE(
+            L"SQLColAttributeW returns SQL_SUCCESS, *CharacterAttributePtr = %s\n",
+            static_cast<SQLWCHAR*>(CharacterAttributePtr)
+            );
```
Comment (ityshko) [inline CdapOdbc/core.cpp]: closing brace must have the same indent as a statement

File: CdapOdbc/core.cpp [hunk 18]
```diff
@@ -699,13 +737,141 @@ SQLRETURN SQL_API SQLColAttributeW(
...
+          return SQL_SUCCESS;
+        }
+      case SQL_COLUMN_LABEL:
+        if (CharacterAttributePtr) {
+          Argument::fromStdString(
+            columnInfo->getShortName(),
+            static_cast<SQLWCHAR*>(CharacterAttributePtr),
+            BufferLength,
+            StringLengthPtr
+            );
```
Comment (ityshko) [inline CdapOdbc/core.cpp]: closing brace must have the same indent as a statement

File: CdapOdbc/core.cpp [hunk 19]
```diff
@@ -699,13 +737,141 @@ SQLRETURN SQL_API SQLColAttributeW(
...
+          TRACE(
+            L"SQLColAttributeW returns SQL_SUCCESS, *CharacterAttributePtr = %s\n",
+            static_cast<SQLWCHAR*>(CharacterAttributePtr)
+            );
```
Comment (ityshko) [inline CdapOdbc/core.cpp]: closing brace must have the same indent as a statement

File: CdapOdbc/core.cpp [hunk 20]
```diff
@@ -699,13 +737,141 @@ SQLRETURN SQL_API SQLColAttributeW(
...
+
+      /* Column extended descriptor fields */
+      case SQL_DESC_BASE_COLUMN_NAME:
+        if (CharacterAttributePtr) {
+          Argument::fromStdString(
+            columnInfo->getShortName(),
+            static_cast<SQLWCHAR*>(CharacterAttributePtr),
+            BufferLength,
+            StringLengthPtr
+            );
```
Comment (ityshko) [inline CdapOdbc/core.cpp]: closing brace must have the same indent as a statement

File: CdapOdbc/core.cpp [hunk 21]
```diff
@@ -699,13 +737,141 @@ SQLRETURN SQL_API SQLColAttributeW(
...
+        if (CharacterAttributePtr) {
+          Argument::fromStdString(
+            columnInfo->getShortName(),
+            static_cast<SQLWCHAR*>(CharacterAttributePtr),
+            BufferLength,
+            StringLengthPtr
+          );
+          TRACE(L"SQLColAttributeW returns SQL_SUCCESS, *CharacterAttributePtr = %s\n",
+            static_cast<SQLWCHAR*>(CharacterAttributePtr)
+            );
```
Comment (ityshko) [inline CdapOdbc/core.cpp]: closing brace must have the same indent as a statement

**Pull request merged**